### PR TITLE
Change `PUBLIC` to `PRIVATE` in `target_compile_options`

### PR DESCRIPTION
**Pull request by zsef123**
About this issues, https://github.com/pytorch/pytorch/issues/31283, https://github.com/pytorch/vision/issues/2001

When build includes `Torch` packages, gcc arguments are injected in `nvcc` and nvcc occurs fatal error like this
```
nvcc fatal   : Unknown option 'Wall'
```

## To reproduce
main.cu
```
#include <stdio.h>

int main() {
    printf("Hello World\n");
    return 0;
}
```
CMakeLists.txt
```
cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
project(test_build)

set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${TORCH_CXX_FLAGS}")

enable_language(CUDA)
add_definitions(-DCUDA_HAS_FP16=1)
add_definitions(-D__CUDA_NO_HALF_OPERATORS__)
add_definitions(-D__CUDA_NO_HALF_CONVERSIONS__)
add_definitions(-D__CUDA_NO_HALF2_OPERATORS__)

find_package(Torch REQUIRED)

add_executable(o.o main.cu)
target_link_libraries(o.o PRIVATE ${TORCH_LIBRARIES})

```
Running the scripts,
```
➜ cmake -DCMAKE_PREFIX_PATH="libtorch" -DCMAKE_EXPORT_COMPILE_COMMANDS=ON ..
-- The C compiler identification is GNU 8.3.0
-- The CXX compiler identification is GNU 8.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The CUDA compiler identification is NVIDIA 10.1.243
-- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc
-- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc - works
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found CUDA: /usr/local/cuda (found version "10.1") 
-- Caffe2: CUDA detected: 10.1
-- Caffe2: CUDA nvcc is: /usr/local/cuda/bin/nvcc
-- Caffe2: CUDA toolkit directory: /usr/local/cuda
-- Caffe2: Header version is: 10.1
-- Found CUDNN: /usr/local/cuda/lib64/libcudnn.so  
-- Found cuDNN: v7.6.4  (include: /usr/local/cuda/include, library: /usr/local/cuda/lib64/libcudnn.so)
-- Autodetected CUDA architecture(s):  7.0 7.0 7.0 7.0
-- Added CUDA NVCC flags for: -gencode;arch=compute_70,code=sm_70
-- Found Torch: /home/user/torch_build/libtorch/lib/libtorch.so  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/user/torch_build/tspine/build

➜ cmake --build . --config Release
Scanning dependencies of target o.o
[ 50%] Building CUDA object CMakeFiles/o.o.dir/main.cu.o
nvcc fatal   : Unknown option 'Wall'
CMakeFiles/o.o.dir/build.make:79: recipe for target 'CMakeFiles/o.o.dir/main.cu.o' failed
make[2]: *** [CMakeFiles/o.o.dir/main.cu.o] Error 1
CMakeFiles/Makefile2:92: recipe for target 'CMakeFiles/o.o.dir/all' failed
make[1]: *** [CMakeFiles/o.o.dir/all] Error 2
Makefile:100: recipe for target 'all' failed
make: *** [all] Error 2

➜ cat compile_commands.json 
[
{
  "directory": "/home/user/torch_build/tspine/build",
  "command": "/usr/local/cuda/bin/nvcc  -DAT_PARALLEL_OPENMP=1 -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF2_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF_OPERATORS__ -isystem=/home/user/torch_build/libtorch/include -isystem=/home/user/torch_build/libtorch/include/torch/csrc/api/include -isystem=/usr/local/cuda/include    -D_GLIBCXX_USE_CXX11_ABI=1 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fopenmp -std=c++14 -x cu -c /home/user/torch_build/tspine/main.cu -o CMakeFiles/o.o.dir/main.cu.o",
  "file": "/home/user/torch_build/tspine/main.cu"
}
]
```

In nvcc, `-Wall` must using with  `-Xcompiler` but it seems flags about `gcc`.

So I change `torch` flags to PRIVATE.

If do this, not anymore happen errors but can't pass on torch compile flags to target package

The other ways, separate compile options by `$<COMPILE_LANGUAGE:...>`.
https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#boolean-compile-language-generator-expression

### CODE REVIEW

**Review by peterjc123 — COMMENTED** [review]
I don't this change is needed. There must be a method to correctly propagate the arguments to the host compiler if you use `enable_language(CUDA)`.

Update:
I think you may resolve the issue by replacing `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${TORCH_CXX_FLAGS}")` to the following code.
```cmake
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${TORCH_CXX_FLAGS}")
set(CUDA_PROPAGATE_HOST_FLAGS OFF)
foreach(f ${TORCH_CXX_FLAGS})
  list(APPEND CMAKE_CUDA_FLAGS -Xcompiler ${f})
endforeach()
```
Alternative, there's a simpler apporach which is not general. But I don't know whether it works with the new CUDA support of CMake.
```cmake
set(CUDA_PROPAGATE_HOST_FLAGS_BLACKLIST "-Werror")
set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -Xcompiler -Werror")
```

**Comment by zsef123** [comment]
@peterjc123 Thanks to comment.

I agreed propagate problem.

Would it be okay to PR if the arguments are passed well and no additional scripts are required for the parent package?

or will you keep in upper suggestions?

**Comment by peterjc123** [comment]
@zsef123 I think the problem is at the cmake side. It would be convenient if they could provide sth like CMAKE_CUDA_HOST_COMPILER_FLAGS.

**Comment by bmanga** [comment]
@peterjc123 you can use generator expressions to guard for device compilers.

**Comment by peterjc123** [comment]
@bmanga Sounds reasonable. Could you please do that?

**Comment by bmanga** [comment]
@peterjc123 I can give it a try over the weekend.

**Comment by zsef123** [comment]
@bmanga
Is this about the `compile_launguage:CUDA`?

**Comment by bmanga** [comment]
@zsef123 I haven't really thought it out yet, but that's what I had in mind yes (or probably checking for C and CXX given that CUDA was introduced later in cmake).

**Comment by bmanga** [comment]
@peterjc123 Just to be completely clear on this, why is torch trying to inject compilation flags into users' cmake targets?

**Comment by peterjc123** [comment]
@ezyang may know this better than me.

**Comment by ezyang** [comment]
I'm having trouble following the discussion. Which injection are you referring to? In general when you define a cmake library, it may specify that certain compilation flags must be used when compiling downstream libraries/executables (e.g., PUBLIC settings).

**Comment by bmanga** [comment]
@ezyang yes, but I can't see how  `-Wall`, `-Wextra` and friends are necessary to use `torch_cpu` (I would say that they should not be in the cmake at all, but definitely not forced upon target consumers).

Of all those added publicly to the target, I think only the openmp ones are necessary.

**Comment by zsef123** [comment]
@ezyang 
Problem is about gcc and nvcc. 

The flag related with `torch_cpu` like `-Wall -Wextra` for gcc has injected to the nvcc.

So nvcc compiler error on the building custom project included torch.

In my case, now I have ignore all compile flag on torch

**Comment by bmanga** [comment]
@zsef123 I actually think that this PR wasn't that unreasonable. I don't see the need for those compile options being propagated.

**Comment by ezyang** [comment]
I do agree that we should never be injecting `-Wall -Wextra` into user builds. Wall is never appropriate for an end user build, it should only be used by devs. I'm not too sure what the best way of enforcing this is, however.

**Comment by zsef123** [comment]
I thought the Pytorch developers want the build flags injecting to project inlclued torch.

but if not, then more options can be considered.

**Comment by peterjc123** [comment]
I guess we still need some of them because we want to make the user build warning clean. But the core problem here is how to properly propagate them with and without the new language support of cuda in CMake.

**Comment by bmanga** [comment]
@peterjc123 Forcing those flags on the user build is not the correct approach though. If all libraries started adding flags to my target I would have no idea what flags I end up building with, and those libraries most likely would end up stepping on each others' feet (eg. libA adds `-Wmaybe-uninitialized` and libB adds `-Wno-maybe-uninitialized`). That may also lead to an implicit dependency with the order of target inclusion.

If we don't want to generate warnings when torch headers are used in user builds the two IMO reasonable options are:
* Guard the offending code in headers with diagnostic pragmas  (eg. `#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"`).
* Set the torch headers as SYSTEM headers in targets meant to be consumed by users, as those do not generate warnings.

**Comment by peterjc123** [comment]
@bmanga This may be a point of improvement.

**Comment by lbuchy** [comment]
@bmanga Just chiming in here.  Option 2 that you presented would fix the issue of ensuring that warnings are not generated when a user includes the torch provided cmake targets.
However, I don't think it fully addresses the original issue.  Ie. it is currently not possible to compile CUDA source files when using `target_link_libraries(my_cuda_target PRIVATE torch)` using PyTorch 1.5+.

I think you hit the nail on the head with your example with `-Wmaybe-uninitialized` vs `-Wno-maybe-initialized`.  It isn't this library's place to force the user to use certain warning flags.  Some may argue it isn't even torch's place to suppress warnings caused by the inclusion of headers. Those warnings may actually point to problems that should be addressed in the headers themselves, this is useful information that torch would be suppressing. 

If a user wants to add a system library and suppress the warnings on its headers, they are free to do so in their own build system.  Granted, it is a little more complex than `target_link_library(my_lib PRIVATE torch)`.  They would need to pull the INTERFACE_INCLUDE_DIR property off of `torch` and set the INTERFACE_SYSTEM_INCLUDE_DIR property on their target.
Something like what is suggested in https://stackoverflow.com/questions/52135983/cmake-target-link-libraries-include-as-system-to-suppress-compiler-warnings

tl;dr : We leave it as a user's decision to suppress warnings from torch headers.  We don't impose flags onto the user's targets by privatizing the compiler options (doing so breaks things in unexpected ways).

**Comment by bmanga** [comment]
@lbuchy I think you misunderstood my point. I do think those compile options (with maybe the exception of openmp, not sure about that one) should be PRIVATE.

However, I also think that libtorch shouldn't generate warnings in user code. This is because users have no way of fixing the code that generates them (well, they can, but that's essentially forking the library) and they just become noise.
I don't see the value in having users find a way to disable the warnings, because I can't think of a reason they'd want to see them in the first place.

While the ideal outcome would be to have the library be warnings-free in all configurations, it's also a big undertaking to make sure it does. Those I mentioned are two possibilities to set the compile options to be private but still keep the library warnings-free for users.

**Comment by lbuchy** [comment]
@bmanga Sorry for misunderstanding, just wanted to clarify I guess.  As a bit more evidence that making those PRIVATE is the right thing, it is in line with [Danel Pfeifer's "Effective CMake" talk](https://github.com/boostcon/cppnow_presentations_2017/blob/master/05-19-2017_friday/effective_cmake__daniel_pfeifer__cppnow_05-19-2017.pdf). Slide 20.  See [relevant section of talk](https://youtu.be/bsXLMQ6WgIk?t=1755)

Just took a look at adding SYSTEM to the `target_include_directories` of torch_cpu, torch_cuda and torch_hip.  One concern about this is doing so would treat the include directories as SYSTEM even internally which would suppress warnings while building torch. Definitely not desirable.

I agree that the warnings are noise and it would be great to disable them for users without their intervention.  Too bad cmake doesn't have support for `target_link_libraries(my_target SYSTEM PRIVATE third_party_lib)`.  I also agree that the ultimate solution is to just fix pytorch's headers, however large of an undertaking that is.  

My concern is that adding SYSTEM will basically lock the user into silencing the warnings by default. Some extra effort would need to be taken to unsilence the warnings.  I would venture a guess that pytorch would get more PRs that fix the warnings in its headers if they are seen by default and silenced by the user if they don't care about them.

I'd be interested in knowing what other third party libraries do when declaring target_include_directories.  A brief glance around at some libraries randomly pulled from the (awesome-cpp repo)[https://github.com/fffaraz/awesome-cpp] show the following do not add SYSTEM to their INTERFACE_INCLUDE_DIR properties
* (Protobuffers)[https://github.com/protocolbuffers/protobuf/blob/master/cmake/libprotobuf.cmake] (the cmake build system, not bazel)
* (spdlog)[https://github.com/gabime/spdlog/blob/v1.x/CMakeLists.txt]
* (opencv)[https://github.com/opencv/opencv/blob/master/cmake/templates/OpenCVConfig.cmake.in] (looks like data cmake build system, sets INTERFACE_INCLUDE_DIRECTORIES directly)
* (Eigen)[https://gitlab.com/libeigen/eigen/-/blob/master/cmake/FindEigen3.cmake]

Not sure though if that is necessarily what torch _should_ do - all of these libraries may have already cleaned up their headers. Just wanted to find some example usage of the SYSTEM flag in a mature third party library.

**Comment by bmanga** [comment]
@lbuchy no worries :) I just wanted to clarify my stand as well.

See #38914 to how I plan to only silence the warnings for user code.

I still disagree on the usefulness of warnings for users. I assume there will be an increase of issues being created for them, and potentially some PRs about it. But this will be to great detriment for everyone, especially for those building with `-Werror` (myself included).

I think the point about those libraries is that they never hid the warnings, so they had a constant stream of fixes to address them. On the other hand, torch has been hiding them, so I'm not sure how smooth it would be for users to suddenly get hundreds after an update.
We could argue if that was the best choice, but there definitely is appeal for having warnings that only concern the code you wrote and can do something about.

Anyways for an example of SYSTEM headers being used for production code you cant take a look at  the [conan](https://conan.io/) package manager. When including a package, all the libraries's headers are implicitly set as SYSTEM.
